runtime/src/bank.rs

```diff
@@ -4788,86 +4796,114 @@ impl Bank {
-        instruction_context.configure(if is_upgradeable_loader { &[0, 1] } else { &[0] }, &[], &[]);
-        transaction_context.push().unwrap();
-        let instruction_context = transaction_context
-            .get_current_instruction_context()
-            .unwrap();
-        let program = instruction_context
-            .try_borrow_program_account(&transaction_context, 0)
-            .unwrap();
-        let programdata = if is_upgradeable_loader {
-            Some(
-                instruction_context
-                    .try_borrow_program_account(&transaction_context, 1)
-                    .unwrap(),
-            )
-        } else {
-            None
-        };
+        ProgramAccountLoadResult::InvalidAccountData
+    }
+
+    pub fn load_program(&self, pubkey: &Pubkey) -> Arc<LoadedProgram> {
         let program_runtime_environment_v1 = self
             .loaded_programs_cache
             .read()
             .unwrap()
             .program_runtime_environment_v1
             .clone();
-        solana_bpf_loader_program::load_program_from_account(
-            &self.feature_set,
-            None, // log_collector
-            &program,
-            programdata.as_ref().unwrap_or(&program),
-            program_runtime_environment_v1.clone(),
-        )
-        .map(|(loaded_program, metrics)| {
-            let mut timings = ExecuteDetailsTimings::default();
-            metrics.submit_datapoint(&mut timings);
-            loaded_program
-        })
+
+        let mut load_program_metrics = LoadProgramMetrics {
+            program_id: pubkey.to_string(),
+            ..LoadProgramMetrics::default()
+        };
+
+        let loaded_program = match self.load_program_accounts(pubkey) {
+            ProgramAccountLoadResult::AccountNotFound => Ok(LoadedProgram::new_tombstone(
+                self.slot,
+                LoadedProgramType::Closed,
+            )),
+
+            ProgramAccountLoadResult::InvalidAccountData => {
+                Err(InstructionError::InvalidAccountData)
+            }
+
+            ProgramAccountLoadResult::ProgramOfLoaderV1orV2(program_account) => {
+                solana_bpf_loader_program::load_program_from_bytes(
+                    &self.feature_set,
+                    None,
+                    &mut load_program_metrics,
+                    program_account.data(),
+                    program_account.owner(),
+                    program_account.data().len(),
+                    0,
+                    program_runtime_environment_v1.clone(),
+                )
+            }
+
+            ProgramAccountLoadResult::ProgramOfLoaderV3(
+                program_account,
+                programdata_account,
+                slot,
+            ) => programdata_account
+                .data()
+                .get(UpgradeableLoaderState::size_of_programdata_metadata()..)
+                .ok_or(InstructionError::InvalidAccountData)
```

Comment:
This originally resulted in `InstructionError::AccountDataTooSmall` in `load_program_from_account()` and all errors from there were turned into `LoadedProgramType::FailedVerification` in `load_program()`.

Reply:
`LoadedProgramType::FailedVerification` resulted in a tombstone. That's what the `unwrap_or_else` clause at line 4897 is achieving in this PR.